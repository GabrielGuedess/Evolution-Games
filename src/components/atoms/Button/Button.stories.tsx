import { Story, Meta } from '@storybook/react/types-6-0';
import { Funnel } from 'phosphor-react';

import Button, { ButtonProps } from './Button';

export default {
  title: 'Atoms/Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string',
    },
    size: {
      control: {
        type: 'select',
        options: ['small', 'medium', 'large'],
      },
    },
    variant: {
      control: {
        type: 'select',
        options: ['solid', 'outline'],
      },
    },
  },
  args: {
    variant: 'solid',
  },
} as Meta<ButtonProps>;

export const Default: Story<ButtonProps> = args => <Button {...args} />;

Default.args = {
  children: 'Buy now',
};

export const Disable: Story<ButtonProps> = args => <Button {...args} />;

Disable.args = {
  children: 'Buy now',
  disabled: true,
};

export const fullWidth: Story<ButtonProps> = args => <Button {...args} />;

fullWidth.args = {
  children: 'Buy now',
  fullWidth: true,
};
export const withIcon: Story<ButtonProps> = args => <Button {...args} />;

withIcon.args = {
  size: 'medium',
  children: 'Buy now',
  icon: <Funnel weight="fill" size={20} className="icon" />,
};
export const OnlyIcon: Story<ButtonProps> = args => <Button {...args} />;

OnlyIcon.args = {
  size: 'medium',
  icon: <Funnel weight="fill" size={20} className="icon" />,
};

export const asLink: Story<ButtonProps> = args => <Button {...args} />;

asLink.args = {
  size: 'large',
  children: 'Buy now',
  as: 'a',
  href: '/link',
};

export const outline: Story<ButtonProps> = args => <Button {...args} />;

outline.args = {
  children: 'All',
  variant: 'outline',
};

export const minimal: Story<ButtonProps> = args => <Button {...args} />;

minimal.args = {
  icon: <Funnel weight="fill" size={20} className="icon" />,
  minimal: true,
};
