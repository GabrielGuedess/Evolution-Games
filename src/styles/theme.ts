export default {
  grid: {
    container: '165.6rem',
  },
  border: {
    radius: {
      regular: '1rem',
    },
  },
  font: {
    family: {
      primary:
        "'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    },
    regular: 400,
    medium: 500,
    sizes: {
      xxsmall: '1rem',
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      xmedium: '1.8rem',
      xxmedium: '2.0rem',
      big: '2.2rem',
      xbig: '2.5rem',
      xxbig: '2.8rem',
      xxxbig: '3.0rem',
      large: '3.2rem',
      xlarge: '4.5rem',
      xxlarge: '5.0rem',
      extra: '6.0rem',
    },
  },
  colors: {
    primary: '#7C3AED',
    secondary: '#717B8A',
    mainBg: '#04040C',
    white: '#FFFFFF',
    whiteText: '#F4F4F5',
    gameDetails: '#010914',
    gameGreat: '#0FBF00',
    gameAverage: '#FFF720',
    gameBad: '#FF3333',
  },
  spacings: {
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem',
    big: '5.8rem',
    huge: '8.0rem',
    section: '10rem',
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50,
  },
  shadows: {
    text: '0 0.8rem 1rem rgba(118, 74, 241, 0.5)',
    box: '0 0.8rem 2.8rem rgba(118, 74, 241, 0.5)',
  },
} as const;
