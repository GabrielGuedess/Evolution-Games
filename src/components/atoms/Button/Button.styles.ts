import media from 'styled-media-query';

import styled, { css, DefaultTheme } from 'styled-components';

import { ButtonProps } from './Button';

export type WrapperProps = {
  hasIcon: boolean;
} & Pick<ButtonProps, 'size' | 'fullWidth' | 'variant' | 'minimal'>;

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.small};
  `,

  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,

  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,

  fullWidth: () => css`
    width: 100%;
  `,

  withIcon: (theme: DefaultTheme) => css`
    span {
      margin-left: ${theme.spacings.xsmall};
    }
  `,

  solid: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.whiteText};

    :not(:disabled):is(:hover, :focus) {
      box-shadow: ${theme.shadows.box};
      color: ${theme.colors.whiteText};
    }

    .effect {
      fill: ${theme.colors.primary};
    }
  `,

  outline: (theme: DefaultTheme) => css`
    background: transparent;
    color: ${theme.colors.secondary};

    .effect {
      fill: none;
      stroke-width: 2px;
      stroke: ${theme.colors.secondary};

      path {
        transition: 0.3s cubic-bezier(0.68, -0.6, 0.32, 1.6);
      }
    }

    :not(:disabled):is(:hover, :focus) .effect {
      stroke: ${theme.colors.primary};
    }

    :not(:disabled):is(:hover, :focus) {
      color: ${theme.colors.primary};
    }
  `,

  minimal: (theme: DefaultTheme) => css`
    background: transparent;
    color: ${theme.colors.secondary};

    :not(:disabled):is(:hover, :focus) {
      box-shadow: 0 0 0 transparent;
      color: ${theme.colors.whiteText};
    }

    .effect {
      display: none;
    }
  `,

  disabled: () => css`
    :disabled {
      cursor: not-allowed;
      filter: saturate(30%);
    }
  `,
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, hasIcon, size, fullWidth, disabled, variant, minimal }) => css`
    width: fit-content;
    cursor: pointer;
    position: relative;
    outline: none;
    display: flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    border-radius: 5px;
    transition: all 0.3s cubic-bezier(0.68, -0.6, 0.32, 1.6);

    ${!!size && wrapperModifiers[size](theme)};
    ${!!variant && wrapperModifiers[variant](theme)}
    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
    ${minimal && wrapperModifiers.minimal(theme)}
    ${!!fullWidth && wrapperModifiers.fullWidth()};
    ${disabled && wrapperModifiers.disabled()};

    :not(:disabled):is(:hover, :focus) {
      filter: brightness(1.1);
    }

    :not(:disabled):active {
      filter: brightness(0.9);
    }

    span {
      font-size: ${theme.font.sizes.medium};
      font-weight: 500;
      line-height: 1;
      display: block;
      position: relative;
      transition: transform 0.3s cubic-bezier(0.68, -0.6, 0.32, 1.6);
      z-index: 1;
    }

    :not(:disabled):is(:hover, :focus) span {
      transform: scale(1.05);
    }

    :not(:disabled):is(:hover, :focus) .icon {
      transform: scale(1.2);
    }

    :not(:disabled):active span {
      transform: scale(0.95);
    }

    svg {
      transition: transform 0.3s cubic-bezier(0.68, -0.6, 0.32, 1.6);
    }

    .effect {
      position: absolute;
      top: initial;
      left: initial;
      width: 100%;
      height: 100%;
      z-index: -1;

      path {
        transition: 0.3s cubic-bezier(0.68, -0.6, 0.32, 1.6);
      }

      ${media.greaterThan('small')`
        top: -5%;
        left: -5%;
        width: 110%;
        height: 110%;
      `}
    }

    :not(:disabled):is(:hover, :focus) .effect path {
      d: path(
        'M0,0 C0,-5 100,-5 100,0 C105,0 105,100 100,100 C100,105 0,105 0,100 C-5,100 -5,0 0,0'
      );
    }

    :not(:disabled):active .effect path {
      d: path(
        'M0 5C0 2.23858 2.23858 0 5 0H95C97.7614 0 100 2.23858 100 5V95C100 97.7614 97.7614 100 95 100H5C2.23858 100 0 97.7614 0 95V5Z'
      );
    }
  `}
`;
