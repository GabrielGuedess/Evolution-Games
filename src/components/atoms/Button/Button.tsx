import { AnchorHTMLAttributes, ButtonHTMLAttributes, forwardRef } from 'react';

import * as S from './Button.styles';

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large';
  fullWidth?: boolean;
  icon?: JSX.Element;
  children: React.ReactNode;
  variant?: 'solid' | 'outline';
  as?: React.ElementType;
} & ButtonTypes;

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  { children, icon, size = 'medium', variant = 'solid', fullWidth, ...props },
  ref,
) => (
  <S.Wrapper
    hasIcon={!!icon}
    size={size}
    ref={ref}
    fullWidth={fullWidth}
    variant={variant}
    aria-label="Button Action"
    {...props}
  >
    {icon}
    {!!children && <span>{children}</span>}

    <svg
      className="effect"
      viewBox="-5 -5 110 110"
      preserveAspectRatio="none"
      aria-hidden="true"
    >
      <path d="M0 8C0 3.58172 3.58172 0 8 0H92C96.4183 0 100 3.58172 100 8V92C100 96.4183 96.4183 100 92 100H8C3.58172 100 0 96.4183 0 92V8Z" />
    </svg>
  </S.Wrapper>
);

export default forwardRef(Button);
