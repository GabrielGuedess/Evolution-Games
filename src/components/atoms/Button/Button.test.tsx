import { Funnel } from 'phosphor-react';
import { renderWithTheme } from 'utils/tests/helpers';

import { screen } from '@testing-library/react';

import theme from 'styles/theme';

import Button from './Button';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    // Arrange
    const { container } = renderWithTheme(<Button>Buy now</Button>);

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveStyle({
      fontSize: theme.font.sizes.small,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.medium}`,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Buy now</Button>);

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveStyle({
      fontSize: theme.font.sizes.xsmall,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.small}`,
    });
  });

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Buy now</Button>);

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveStyle({
      fontSize: theme.font.sizes.medium,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.xlarge}`,
    });
  });

  it('should render a fullWidth version', () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>);

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveStyle({
      width: '100%',
    });
  });

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<Funnel data-testid="icon" />}>Buy now</Button>,
    );

    // Assert
    expect(screen.getByText(/buy now/i)).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should render a disabled Button', () => {
    renderWithTheme(<Button disabled>Buy now</Button>);

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveStyleRule(
      'cursor',
      'not-allowed',
      {
        modifier: ':disabled',
      },
    );
  });

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>,
    );

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveAttribute(
      'href',
      '/link',
    );
  });

  it('should render Button as minimal', () => {
    renderWithTheme(<Button minimal>Buy now</Button>);

    // Assert
    expect(screen.getByLabelText('Button Action')).toHaveStyle({
      background: 'transparent',
      color: theme.colors.secondary,
    });
  });
});
