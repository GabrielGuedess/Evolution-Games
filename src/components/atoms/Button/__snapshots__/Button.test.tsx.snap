// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<Button /> should render the medium size by default 1'] = `
.c0 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  background: #6200EE;
  cursor: pointer;
  position: relative;
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #F4F4F5;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: all 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:not(:disabled):is(:hover,:focus) {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
  box-shadow: 0 0.8rem 2.8rem rgba(118,74,241,0.5);
}

.c0:not(:disabled):active {
  -webkit-filter: brightness(0.9);
  filter: brightness(0.9);
}

.c0 span {
  font-size: 1.6rem;
  font-weight: 500;
  line-height: 1;
  display: block;
  position: relative;
  -webkit-transition: -webkit-transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  -webkit-transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  z-index: 1;
}

.c0:not(:disabled):is(:hover,:focus) span {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c0:not(:disabled):is(:hover,:focus) .icon {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c0:not(:disabled):active span {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
}

.c0 svg {
  -webkit-transition: -webkit-transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  -webkit-transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
}

.c0 .effect {
  fill: #855aff;
  position: absolute;
  top: -5%;
  left: -5%;
  width: 110%;
  height: 110%;
  z-index: -1;
}

.c0 .effect path {
  -webkit-transition: 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
}

.c0:not(:disabled):is(:hover,:focus) .effect path {
  d: path( 'M0,0 C0,-5 100,-5 100,0 C105,0 105,100 100,100 C100,105 0,105 0,100 C-5,100 -5,0 0,0' );
}

.c0:not(:disabled):active .effect path {
  d: path( 'M0,0 C30,10 70,10 100,0 C95,30 95,70 100,100 C70,90 30,90 0,100 C5,70 5,30 0,0' );
}

<button
  aria-label="Button Action"
  class="c0"
>
  <span>
    Buy now
  </span>
  <svg
    aria-hidden="true"
    class="effect"
    preserveAspectRatio="none"
    viewBox="-5 -5 110 110"
  >
    <path
      d="M0,0 C0,0 100,0 100,0 C100,0 100,100 100,100 C100,100 0,100 0,100 C0,100 0,0 0,0"
    />
  </svg>
</button>
`;
